data/lua/carryover_gold.lua

```diff
@@ -0,0 +1,146 @@
+
+	if has_next_scenario then
+		for __, side in ipairs(wesnoth.sides) do
+			if (not side.persistent) or side.lost then
+				goto continue
+			end
+			local finishing_bonus_per_turn = num_villages * side.village_gold + side.base_income
+			local finishing_bonus = side.carryover_bonus * finishing_bonus_per_turn * turns_left
+
+			side.carryover_gold = math.ceil((side.gold + finishing_bonus) * side.carryover_percentage / 100)
+
+			if (not side.is_local) or (not side.controller == "human") then
+				goto continue
+			end
+
+			if num_persistent_teams > 1 then
+				report = report .. "\n\n<b>" .. side.side_name .. "</b>"
+			end
+
+			report = report .. "<small>\n" .. _("Remaining gold: ") .. half_signed_value(side.gold) .. "</small>"
+
+			if side.carryover_bonus ~= 0 then
+				if turns_left > -1 then
+					report = report .. "\n\n<b>" ..
+						_("Turns finished early: ") .. turns_left .. "</b>\n" .. "<small>" ..
+						_("Early finish bonus: ") .. finishing_bonus_per_turn .. _(" per turn") .. "</small>\n" .. "<small>" ..
+						_("Total bonus: ") .. finishing_bonus .. "</small>\n"
+				end
+
+				report = report .. "<small>" .. _("Total gold: ") .. half_signed_value(side.gold + finishing_bonus) .. "</small>"
+			end
+
+			if side.gold > 0 then
+				report = report .. "\n<small>" .. _("Carryover percentage: ") .. side.carryover_percentage .. "</small>"
+			end
+
+			if side.carryover_add then
+				report = report .. "\n\n<big><b>" .. _("Bonus gold: ") .. half_signed_value(side.carryover_gold) .. "</b></big>"
+			else
+				report = report .. "\n\n<big><b>" .. _("Retained gold: ") .. half_signed_value(side.carryover_gold) .. "</b></big>"
+			end
+
+			local goldmsg = ""
+			-- Note that both strings are the same in English, but some languages will
+			-- want to translate them differently.
+			if side.carryover_add then
+				if side.carryover_gold > 0 then
+					goldmsg = _("You will start the next scenario with $gold on top of the defined minimum starting gold.",
+						"You will start the next scenario with $gold on top of the defined minimum starting gold.", side.carryover_gold)
+
+				else
+					goldmsg = _("You will start the next scenario with the defined minimum starting gold.",
+						"You will start the next scenario with the defined minimum starting gold.", side.carryover_gold)
+				end
+			else
+				goldmsg = _(
+					"You will start the next scenario with $gold or its defined minimum starting gold, whichever is higher.",
+					"You will start the next scenario with $gold or its defined minimum starting gold, whichever is higher.",
+					side.carryover_gold)
+			end
+
+			-- xgettext:no-c-format
+			report = report .. "\n" .. goldmsg:vformat{ gold = tostring(side.carryover_gold) }
+
+			::continue::
+		end
+	end
+
+	if show_report then
+		gui.show_popup(title, report)
+	end
+end
+
+
+wesnoth.game_events.add {
+	name = "scenario_end",
+	id = "carryover_gold",
+	first_time_only = false,
+	--priority = -1000,
+	action = function()
```

Comment:
Just `action = do_carryover_gold` should suffice, right?

Reply:
i was considering making the variable `do_carryover_gold` non local, so that it could be replaced by umc. That's why i added this indirection

Reply:
Hmm… someone wanting to replace it entirely could just unregister this event. I guess what you say makes sense if they want to mildly alter the behaviour, maybe…

Reply:
hmm i also wasn't sure how cleanly unregistering an event handler that was set in preload (and by preload i also mean toplevel lua) works, but i guess you just have to unregister them at preload aswell.